lib/ui/painting.dart

```diff
@@ -4451,6 +4433,103 @@ class Shadow {
 }
 
+/// A handle to a read-only byte buffer that is managed by the engine.
+class ImmutableBuffer extends NativeFieldWrapperClass2 {
+  ImmutableBuffer._(this.length);
+
+  /// Creates a copy of the data from a [Uint8List] suitable for internal use
+  /// in the engine.
+  static Future<ImmutableBuffer> fromUint8List(Uint8List list) {
+    final ImmutableBuffer instance = ImmutableBuffer._(list.length);
+    return _futurize((_Callback<void> callback) {
+      instance._init(list, callback);
+    }).then((_) => instance);
+  }
+  void _init(Uint8List list, _Callback<void> callback) native 'ImmutableBuffer_init';
+
+  /// The length, in bytes, of the underlying data.
+  final int length;
+
+  /// Retreives a single byte from the underlying data.
+  int operator [](int index) => elementAt(index);
+
+  /// Retreives a single byte from the underlying data.
+  int elementAt(int index) {
+    if (index >= length || index < 0) {
+      throw RangeError.index(index, this);
+    }
+    return _elementAt(index);
+  }
+  int _elementAt(int index) native 'ImmutableBuffer_elementAt';
+
+  /// Release the resources used by this object. The object is no longer usable
+  /// after this method is called.
+  void dispose() native 'ImmutableBuffer_dispose';
+}
+
+/// A descriptor of data that can be turned into an [Image] via a [Codec].
+///
+/// Use this class to determine the height, width, and byte size of image data
+/// before decoding it.
+class ImageDescriptor extends NativeFieldWrapperClass2 {
+  ImageDescriptor._();
+
+  /// Creates an image descriptor from encoded data in a supported format.
+  static Future<ImageDescriptor> encoded(ImmutableBuffer buffer) {
+    final ImageDescriptor descriptor = ImageDescriptor._();
+    return _futurize((_Callback<void> callback) {
+      descriptor._initEncoded(buffer, callback);
+    }).then((_) => descriptor);
+  }
+  String? _initEncoded(ImmutableBuffer buffer, _Callback<void> callback) native 'ImageDescriptor_initEncoded';
+
+  /// Creates an image descriptor from raw image pixels.
+  ///
+  /// The `pixels` parameter is the pixel data in the encoding described by
+  /// `format`.
+  ///
+  /// The `rowBytes` parameter is the number of bytes consumed by each row of
+  /// pixels in the data buffer. If unspecified, it defaults to `width` multiplied
+  /// by the number of bytes per pixel in the provided `format`.
+  // Not async because there's no expensive work to do here.
+  ImageDescriptor.raw(
+    ImmutableBuffer buffer, {
+    required int width,
+    required int height,
+    int? rowBytes,
+    required PixelFormat pixelFormat,
+  }) {
+    _initRaw(this, buffer, width, height, rowBytes ?? -1, pixelFormat.index);
+  }
+  void _initRaw(ImageDescriptor outDescriptor, ImmutableBuffer buffer, int width, int height, int rowBytes, int pixelFormat) native 'ImageDescriptor_initRaw';
+
+  /// The width, in pixels, of the image.
+  int get width native 'ImageDescriptor_width';
```

Comment:
Can `width`, `height`, or `bytesPerPixel` change? If not, then you can cache them in Dart fields, which will have faster access than a runtime call.
```dart
int? _width;
int get width => _width ??= _getWidth();
int _getWidth() native 'ImageDescriptor_width';
```

Reply:
These are immutable - we can cache them.

Do we have any benchmarks demonstrating this is faster though? At one point I thought I measured this and it didn't make a big difference.

Reply:
Either way, done.

Reply:
I doubt we have any benchmarks doing an A/B comparison, but in general the difference is a null check + load (3-4 instructions) vs. saving live registers, doing the runtime call, looking up the native call by string, setting up handle and allocation arenas in the runtime, etc., etc.., and then popping back out undoing all that. E.g.:

https://github.com/dart-lang/sdk/blob/master/runtime/vm/compiler/assembler/assembler_arm64.cc#L1488,
https://github.com/dart-lang/sdk/blob/master/runtime/vm/compiler/stub_code_compiler_arm64.cc#L74

and so on.